cal/cal/static/js/app.js

```diff
@@ -3,32 +3,120 @@ var analyticsApp = angular.module('analyticsApp', ['nvd3', 'ui.calendar']);
 analyticsApp.controller('LoggedInCtrl', function LoggedInController($scope) {
 });
 
-// Controller to generate graph data from chart directive, cumulative tags
 analyticsApp.controller('TagsCtrl', function($scope, $http){
   var url = '/v1/tags.json';
+  var tagUrl = '/v1/tags/';
+
   $scope.url = url;
   $scope.tags = [];
 
+  // Generate graph data
   $http({ method: 'GET', url: url }).
     success(function successCallback(data) {
-
       for (var i = 0; i < data.results.length; i++) {
         var tag = data.results[i];
         $scope.tags.push({
+          id: tag.id,
           label: tag.label,
           keywords: tag.keywords,
           hours: tag.hours
         });
       }
-
     });
 
+    this.addTag = function(tag) {
+      $http({
+        method: 'POST',
+        url: url,
+        data: $.param({
+          label: tag.label,
+          keywords: tag.keywords,
+          csrfmiddlewaretoken: getCookie('csrftoken')
+        }),
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded'
+        }
+      }).
+        success(function addToList(data) {
+          $scope.tags.push({
+            id: data.id,
+            label: data.label,
+            keywords: data.keywords,
+            hours: data.hours,
+            showEdit: false
+          });
+        });
+    };
+
+    this.editTag = function(tagId) {
+      var tag = search(tagId, $scope.tags);
+      tag.new_label = tag.label;
+      tag.new_keywords = tag.keywords;
+      tag.showEdit = true;
+    };
+
+    this.submit = function(tagId) {
+      var tag = search(tagId, $scope.tags);
+      tag.showEdit = false;
+
+      $http({
+        method: 'POST',
+        url: tagUrl + tagId,
+        data: $.param({
+          label: tag.new_label,
+          keywords: tag.new_keywords,
+          csrfmiddlewaretoken: getCookie('csrftoken'),
+          _method: 'PATCH'
+        }),
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded'
+        }
+      }).
+        success(function editToList(data) {
+          tag.label = data.label;
+          tag.keywords = data.keywords;
+          tag.hours = data.hours;
+        });
+    };
+
+    this.cancel = function(tagId) {
+      var tag = search(tagId, $scope.tags);
+      tag.showEdit = false;
+    };
+
+    this.deleteTag = function(tagId) {
+      $http({
+        method: 'POST',
+        url: tagUrl + tagId,
+        data: $.param({
+          csrfmiddlewaretoken: getCookie('csrftoken'),
+          _method: 'DELETE'
+        }),
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded'
+        }
+      }).
+        success(function removeTagFromList(data) {
+          $scope.tags = $scope.tags.filter(function(tag) {
+            return tag.id !== tagId;
+          });
+        });
+    };
+
+    function search(id, tags){
```

Comment:
Ahh, I should have been more clear. I meant that every time you're calling search, it should be implemented with a `.find(` instead. For example in like 83 it should be:

`var tag = $scope.tags.find(function(tag, index, array) { return element.id == id; });`

Which probably seems verbose since you have to write that everywhere you're using currently using search, but is better than implementing a library function ourselves.

The reason we have to do this is because we haven't finished issue #41 yet, which would change our array accesses to constant time by using a dictionary instead of an array.